File: src/lib/cloud-router/classifier.ts
// ═══════════════════════════════════════════════════════════════
// BUNKER Cloud Router - Frontend Task Classifier
// Provides instant classification feedback in the UI
// Mirrors src-tauri/src/cloud_router/classifier.rs
// ═══════════════════════════════════════════════════════════════

import { TaskCategory, TaskClassification, CloudProvider, CATEGORY_INFO } from './types';

interface CategoryPatterns {
  category: TaskCategory;
  patterns: Array<{ keyword: string; weight: number }>;
}

const PATTERNS: CategoryPatterns[] = [
  {
    category: 'code',
    patterns: [
      { keyword: 'rust', weight: 2.0 },
      { keyword: 'typescript', weight: 2.0 },
      { keyword: 'javascript', weight: 2.0 },
      { keyword: 'python', weight: 2.0 },
      { keyword: 'java', weight: 1.5 },
      { keyword: 'c++', weight: 1.5 },
      { keyword: 'golang', weight: 1.5 },
      { keyword: 'function', weight: 1.5 },
      { keyword: 'class', weight: 1.0 },
      { keyword: 'debug', weight: 2.0 },
      { keyword: 'implement', weight: 1.5 },
      { keyword: 'refactor', weight: 2.0 },
      { keyword: 'optimize', weight: 1.5 },
      { keyword: 'algorithm', weight: 2.0 },
      { keyword: 'data structure', weight: 2.0 },
      { keyword: 'api', weight: 1.5 },
      { keyword: 'endpoint', weight: 1.5 },
      { keyword: 'database', weight: 1.5 },
      { keyword: 'sql', weight: 2.0 },
      { keyword: 'regex', weight: 1.5 },
      { keyword: 'compile', weight: 1.5 },
      { keyword: 'syntax', weight: 1.5 },
      { keyword: 'error', weight: 1.0 },
      { keyword: 'bug', weight: 1.5 },
      { keyword: 'fix', weight: 1.0 },
      { keyword: 'code', weight: 1.5 },
      { keyword: 'programming', weight: 1.5 },
      { keyword: 'script', weight: 1.0 },
      { keyword: 'library', weight: 1.0 },
      { keyword: 'framework', weight: 1.0 },
      { keyword: 'test', weight: 1.0 },
      { keyword: 'unit test', weight: 1.5 },
    ],
  },
  {
    category: 'analysis',
    patterns: [
      { keyword: 'analyze', weight: 2.0 },
      { keyword: 'compare', weight: 1.5 },
      { keyword: 'evaluate', weight: 1.5 },
      { keyword: 'review', weight: 1.5 },
      { keyword: 'examine', weight: 1.5 },
      { keyword: 'assess', weight: 1.5 },
      { keyword: 'breakdown', weight: 1.5 },
      { keyword: 'pros and cons', weight: 2.0 },
      { keyword: 'tradeoffs', weight: 2.0 },
      { keyword: 'trade-offs', weight: 2.0 },
      { keyword: 'advantages', weight: 1.0 },
      { keyword: 'disadvantages', weight: 1.0 },
      { keyword: 'differences', weight: 1.5 },
      { keyword: 'similarities', weight: 1.5 },
    ],
  },
  {
    category: 'reasoning',
    patterns: [
      { keyword: 'why', weight: 1.5 },
      { keyword: 'explain', weight: 1.5 },
      { keyword: 'how does', weight: 1.5 },
      { keyword: 'logic', weight: 2.0 },
      { keyword: 'deduce', weight: 2.0 },
      { keyword: 'because', weight: 1.0 },
      { keyword: 'reason', weight: 1.5 },
      { keyword: 'understand', weight: 1.0 },
      { keyword: 'clarify', weight: 1.0 },
      { keyword: 'what if', weight: 1.5 },
      { keyword: 'implications', weight: 1.5 },
      { keyword: 'consequence', weight: 1.5 },
    ],
  },
  {
    category: 'research',
    patterns: [
      { keyword: 'search', weight: 2.0 },
      { keyword: 'find', weight: 1.5 },
      { keyword: 'latest', weight: 2.0 },
      { keyword: 'current', weight: 1.5 },
      { keyword: 'news', weight: 2.0 },
      { keyword: '2024', weight: 2.0 },
      { keyword: '2025', weight: 2.0 },
      { keyword: '2026', weight: 2.0 },
      { keyword: 'recent', weight: 2.0 },
      { keyword: 'update', weight: 1.0 },
      { keyword: 'today', weight: 1.5 },
      { keyword: 'this week', weight: 2.0 },
      { keyword: 'trending', weight: 2.0 },
      { keyword: 'what happened', weight: 2.0 },
      { keyword: 'who is', weight: 1.5 },
      { keyword: 'where is', weight: 1.5 },
      { keyword: 'when did', weight: 1.5 },
      { keyword: 'statistics', weight: 1.5 },
      { keyword: 'data about', weight: 1.5 },
      { keyword: 'look up', weight: 2.0 },
      { keyword: 'facts about', weight: 1.5 },
    ],
  },
  {
    category: 'creative',
    patterns: [
      { keyword: 'write', weight: 1.5 },
      { keyword: 'story', weight: 2.0 },
      { keyword: 'brainstorm', weight: 2.0 },
      { keyword: 'imagine', weight: 2.0 },
      { keyword: 'creative', weight: 2.0 },
      { keyword: 'ideas', weight: 1.5 },
      { keyword: 'fiction', weight: 2.0 },
      { keyword: 'poem', weight: 2.0 },
      { keyword: 'poetry', weight: 2.0 },
      { keyword: 'narrative', weight: 2.0 },
      { keyword: 'character', weight: 1.5 },
      { keyword: 'plot', weight: 1.5 },
      { keyword: 'dialogue', weight: 1.5 },
      { keyword: 'screenplay', weight: 2.0 },
      { keyword: 'novel', weight: 2.0 },
      { keyword: 'blog', weight: 1.5 },
      { keyword: 'article', weight: 1.0 },
      { keyword: 'content', weight: 1.0 },
      { keyword: 'copywriting', weight: 2.0 },
      { keyword: 'marketing', weight: 1.5 },
      { keyword: 'slogan', weight: 2.0 },
      { keyword: 'tagline', weight: 2.0 },
    ],
  },
  {
    category: 'general',
    patterns: [
      { keyword: 'summarize', weight: 1.5 },
      { keyword: 'translate', weight: 2.0 },
      { keyword: 'list', weight: 1.0 },
      { keyword: 'format', weight: 1.0 },
      { keyword: 'convert', weight: 1.0 },
      { keyword: 'rewrite', weight: 1.0 },
      { keyword: 'simplify', weight: 1.0 },
      { keyword: 'help', weight: 0.5 },
      { keyword: 'what is', weight: 1.0 },
      { keyword: 'define', weight: 1.0 },
    ],
  },
  {
    category: 'image',
    patterns: [
      { keyword: 'image', weight: 2.0 },
      { keyword: 'picture', weight: 2.0 },
      { keyword: 'generate image', weight: 3.0 },
      { keyword: 'create image', weight: 3.0 },
      { keyword: 'dall-e', weight: 3.0 },
      { keyword: 'dalle', weight: 3.0 },
      { keyword: 'draw', weight: 2.0 },
      { keyword: 'illustration', weight: 2.0 },
      { keyword: 'visual', weight: 1.5 },
      { keyword: 'photo', weight: 1.5 },
      { keyword: 'artwork', weight: 2.0 },
    ],
  },
];

/**
 * Get the suggested provider for a category
 */
function getSuggestedProvider(category: TaskCategory): CloudProvider {
  return CATEGORY_INFO[category].suggestedProvider;
}

/**
 * Generate human-readable reasoning for the classification
 */
function generateReasoning(
  category: TaskCategory,
  keywords: string[],
  confidence: number
): string {
  const confidenceLevel =
    confidence > 0.8 ? 'High' : confidence > 0.5 ? 'Medium' : 'Low';

  const categoryDescriptions: Record<TaskCategory, string> = {
    code: 'code-related task',
    analysis: 'analytical task',
    reasoning: 'reasoning/explanation task',
    research: 'research/current information task',
    creative: 'creative writing task',
    general: 'general utility task',
    image: 'image generation task',
  };

  const providerReasons: Record<TaskCategory, string> = {
    code: 'Claude excels at code generation and debugging',
    analysis: 'Claude provides thorough analytical breakdowns',
    reasoning: 'Claude offers detailed logical explanations',
    research: 'Perplexity has real-time web search capabilities',
    creative: 'ChatGPT is strong at creative content generation',
    general: 'ChatGPT handles general tasks efficiently',
    image: 'ChatGPT has DALL-E integration for image generation',
  };

  const categoryDesc = categoryDescriptions[category];
  const providerReason = providerReasons[category];

  if (keywords.length === 0) {
    return `${confidenceLevel} confidence: Classified as ${categoryDesc} (default). ${providerReason}.`;
  }

  const topKeywords = keywords.slice(0, 3).map((k) => `'${k}'`).join(', ');
  return `${confidenceLevel} confidence: Detected ${categoryDesc} based on keywords: ${topKeywords}. ${providerReason}.`;
}

/**
 * Classify a task prompt into a category with confidence score
 * This runs entirely in the frontend for instant feedback
 */
export function classifyTask(prompt: string): TaskClassification {
  const promptLower = prompt.toLowerCase();

  // Initialize scores
  const scores: Record<TaskCategory, number> = {
    code: 0,
    analysis: 0,
    reasoning: 0,
    research: 0,
    creative: 0,
    general: 0,
    image: 0,
  };

  const matchedKeywords: string[] = [];

  // Score each category
  for (const { category, patterns } of PATTERNS) {
    for (const { keyword, weight } of patterns) {
      if (promptLower.includes(keyword)) {
        scores[category] += weight;
        if (!matchedKeywords.includes(keyword)) {
          matchedKeywords.push(keyword);
        }
      }
    }
  }

  // Code block detection (high confidence for code)
  if (
    prompt.includes('```') ||
    prompt.includes('fn ') ||
    prompt.includes('function ') ||
    prompt.includes('def ') ||
    prompt.includes('class ')
  ) {
    scores.code += 3.0;
    if (!matchedKeywords.includes('code_block')) {
      matchedKeywords.push('code_block');
    }
  }

  // Length modifier (longer prompts tend to be more complex)
  if (prompt.length > 500) {
    scores.analysis += 1.0;
    scores.reasoning += 0.5;
  }

  // Question detection
  if (prompt.trim().endsWith('?')) {
    scores.research += 0.5;
  }

  // Find the highest scoring category
  let maxScore = 0;
  let maxCategory: TaskCategory = 'general';

  for (const [category, score] of Object.entries(scores)) {
    if (score > maxScore) {
      maxScore = score;
      maxCategory = category as TaskCategory;
    }
  }

  // Calculate confidence (normalize to 0-1 range, cap at 0.99)
  const totalScore = Object.values(scores).reduce((a, b) => a + b, 0);
  const confidence = totalScore > 0
    ? Math.min(maxScore / totalScore, 0.99)
    : 0.5; // Default confidence if no patterns matched

  // Generate reasoning
  const reasoning = generateReasoning(maxCategory, matchedKeywords, confidence);

  // Sort and dedupe keywords
  const uniqueKeywords = [...new Set(matchedKeywords)].sort();

  return {
    category: maxCategory,
    confidence,
    suggested_provider: getSuggestedProvider(maxCategory),
    reasoning,
    keywords_matched: uniqueKeywords,
  };
}

/**
 * Debounced classifier for use during typing
 */
export function createDebouncedClassifier(
  callback: (result: TaskClassification) => void,
  delay: number = 300
): (prompt: string) => void {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  return (prompt: string) => {
    if (timeoutId) {
      clearTimeout(timeoutId);
    }

    timeoutId = setTimeout(() => {
      const result = classifyTask(prompt);
      callback(result);
    }, delay);
  };
}

File: src-tauri/src/cloud_router/classifier.rs
// Task Classifier
// Rule-based classification for routing tasks to appropriate providers

use crate::cloud_router::types::{CloudProvider, TaskCategory, TaskClassification};
use regex::Regex;

/// Keyword patterns for each category with their weights
struct CategoryPattern {
    category: TaskCategory,
    keywords: Vec<(&'static str, f32)>, // (pattern, weight)
}

/// Classify a task prompt into a category with confidence score
pub fn classify_task(prompt: &str) -> TaskClassification {
    let prompt_lower = prompt.to_lowercase();
    let mut scores: std::collections::HashMap<TaskCategory, f32> = std::collections::HashMap::new();
    let mut matched_keywords: Vec<String> = Vec::new();

    // Initialize all scores to 0
    scores.insert(TaskCategory::Code, 0.0);
    scores.insert(TaskCategory::Analysis, 0.0);
    scores.insert(TaskCategory::Reasoning, 0.0);
    scores.insert(TaskCategory::Research, 0.0);
    scores.insert(TaskCategory::Creative, 0.0);
    scores.insert(TaskCategory::General, 0.0);
    scores.insert(TaskCategory::Image, 0.0);

    // Code patterns
    let code_patterns = [
        ("rust", 2.0),
        ("typescript", 2.0),
        ("javascript", 2.0),
        ("python", 2.0),
        ("java", 1.5),
        ("c++", 1.5),
        ("golang", 1.5),
        ("function", 1.5),
        ("class", 1.0),
        ("debug", 2.0),
        ("implement", 1.5),
        ("refactor", 2.0),
        ("optimize", 1.5),
        ("algorithm", 2.0),
        ("data structure", 2.0),
        ("api", 1.5),
        ("endpoint", 1.5),
        ("database", 1.5),
        ("sql", 2.0),
        ("regex", 1.5),
        ("compile", 1.5),
        ("syntax", 1.5),
        ("error", 1.0),
        ("bug", 1.5),
        ("fix", 1.0),
        ("code", 1.5),
        ("programming", 1.5),
        ("script", 1.0),
        ("library", 1.0),
        ("framework", 1.0),
        ("test", 1.0),
        ("unit test", 1.5),
    ];

    // Analysis patterns
    let analysis_patterns = [
        ("analyze", 2.0),
        ("compare", 1.5),
        ("evaluate", 1.5),
        ("review", 1.5),
        ("examine", 1.5),
        ("assess", 1.5),
        ("breakdown", 1.5),
        ("pros and cons", 2.0),
        ("tradeoffs", 2.0),
        ("trade-offs", 2.0),
        ("advantages", 1.0),
        ("disadvantages", 1.0),
        ("differences", 1.5),
        ("similarities", 1.5),
    ];

    // Reasoning patterns
    let reasoning_patterns = [
        ("why", 1.5),
        ("explain", 1.5),
        ("how does", 1.5),
        ("logic", 2.0),
        ("deduce", 2.0),
        ("because", 1.0),
        ("reason", 1.5),
        ("understand", 1.0),
        ("clarify", 1.0),
        ("what if", 1.5),
        ("implications", 1.5),
        ("consequence", 1.5),
    ];

    // Research patterns
    let research_patterns = [
        ("search", 2.0),
        ("find", 1.5),
        ("latest", 2.0),
        ("current", 1.5),
        ("news", 2.0),
        ("2024", 2.0),
        ("2025", 2.0),
        ("recent", 2.0),
        ("update", 1.0),
        ("today", 1.5),
        ("this week", 2.0),
        ("trending", 2.0),
        ("what happened", 2.0),
        ("who is", 1.5),
        ("where is", 1.5),
        ("when did", 1.5),
        ("statistics", 1.5),
        ("data about", 1.5),
        ("look up", 2.0),
        ("facts about", 1.5),
    ];

    // Creative patterns
    let creative_patterns = [
        ("write", 1.5),
        ("story", 2.0),
        ("brainstorm", 2.0),
        ("imagine", 2.0),
        ("creative", 2.0),
        ("ideas", 1.5),
        ("fiction", 2.0),
        ("poem", 2.0),
        ("poetry", 2.0),
        ("narrative", 2.0),
        ("character", 1.5),
        ("plot", 1.5),
        ("dialogue", 1.5),
        ("screenplay", 2.0),
        ("novel", 2.0),
        ("blog", 1.5),
        ("article", 1.0),
        ("content", 1.0),
        ("copywriting", 2.0),
        ("marketing", 1.5),
        ("slogan", 2.0),
        ("tagline", 2.0),
    ];

    // General patterns
    let general_patterns = [
        ("summarize", 1.5),
        ("translate", 2.0),
        ("list", 1.0),
        ("format", 1.0),
        ("convert", 1.0),
        ("rewrite", 1.0),
        ("simplify", 1.0),
        ("help", 0.5),
        ("what is", 1.0),
        ("define", 1.0),
        ("explain", 0.5), // Lower weight as it overlaps with reasoning
    ];

    // Image patterns
    let image_patterns = [
        ("image", 2.0),
        ("picture", 2.0),
        ("generate image", 3.0),
        ("create image", 3.0),
        ("dall-e", 3.0),
        ("dalle", 3.0),
        ("draw", 2.0),
        ("illustration", 2.0),
        ("visual", 1.5),
        ("photo", 1.5),
        ("artwork", 2.0),
    ];

    // Score each category
    let all_patterns: Vec<(TaskCategory, &[(&str, f32)])> = vec![
        (TaskCategory::Code, &code_patterns),
        (TaskCategory::Analysis, &analysis_patterns),
        (TaskCategory::Reasoning, &reasoning_patterns),
        (TaskCategory::Research, &research_patterns),
        (TaskCategory::Creative, &creative_patterns),
        (TaskCategory::General, &general_patterns),
        (TaskCategory::Image, &image_patterns),
    ];

    for (category, patterns) in all_patterns {
        for (keyword, weight) in patterns.iter() {
            if prompt_lower.contains(keyword) {
                *scores.get_mut(&category).unwrap() += weight;
                matched_keywords.push(keyword.to_string());
            }
        }
    }

    // Code block detection (high confidence for code)
    if prompt.contains("```") || prompt.contains("fn ") || prompt.contains("function ")
        || prompt.contains("def ") || prompt.contains("class ") {
        *scores.get_mut(&TaskCategory::Code).unwrap() += 3.0;
        matched_keywords.push("code_block".to_string());
    }

    // Length modifier (longer prompts tend to be more complex = analysis/reasoning)
    if prompt.len() > 500 {
        *scores.get_mut(&TaskCategory::Analysis).unwrap() += 1.0;
        *scores.get_mut(&TaskCategory::Reasoning).unwrap() += 0.5;
    }

    // Question detection
    if prompt.ends_with('?') {
        *scores.get_mut(&TaskCategory::Research).unwrap() += 0.5;
    }

    // Find the highest scoring category
    let mut max_score = 0.0f32;
    let mut max_category = TaskCategory::General;

    for (category, score) in &scores {
        if *score > max_score {
            max_score = *score;
            max_category = *category;
        }
    }

    // Calculate confidence (normalize to 0-1 range, cap at 0.99)
    let total_score: f32 = scores.values().sum();
    let confidence = if total_score > 0.0 {
        (max_score / total_score).min(0.99)
    } else {
        0.5 // Default confidence if no patterns matched
    };

    // Generate reasoning
    let reasoning = generate_reasoning(max_category, &matched_keywords, confidence);

    // Remove duplicates from matched keywords
    matched_keywords.sort();
    matched_keywords.dedup();

    TaskClassification {
        category: max_category,
        confidence,
        suggested_provider: max_category.suggested_provider(),
        reasoning,
        keywords_matched: matched_keywords,
    }
}

/// Generate human-readable reasoning for the classification
fn generate_reasoning(category: TaskCategory, keywords: &[String], confidence: f32) -> String {
    let confidence_level = if confidence > 0.8 {
        "High"
    } else if confidence > 0.5 {
        "Medium"
    } else {
        "Low"
    };

    let category_desc = match category {
        TaskCategory::Code => "code-related task",
        TaskCategory::Analysis => "analytical task",
        TaskCategory::Reasoning => "reasoning/explanation task",
        TaskCategory::Research => "research/current information task",
        TaskCategory::Creative => "creative writing task",
        TaskCategory::General => "general utility task",
        TaskCategory::Image => "image generation task",
    };

    let provider_reason = match category {
        TaskCategory::Code => "Claude excels at code generation and debugging",
        TaskCategory::Analysis => "Claude provides thorough analytical breakdowns",
        TaskCategory::Reasoning => "Claude offers detailed logical explanations",
        TaskCategory::Research => "Perplexity has real-time web search capabilities",
        TaskCategory::Creative => "ChatGPT is strong at creative content generation",
        TaskCategory::General => "ChatGPT handles general tasks efficiently",
        TaskCategory::Image => "ChatGPT has DALL-E integration for image generation",
    };

    if keywords.is_empty() {
        format!(
            "{} confidence: Classified as {} (default). {}.",
            confidence_level, category_desc, provider_reason
        )
    } else {
        let top_keywords: Vec<&String> = keywords.iter().take(3).collect();
        format!(
            "{} confidence: Detected {} based on keywords: {}. {}.",
            confidence_level,
            category_desc,
            top_keywords.iter().map(|k| format!("'{}'", k)).collect::<Vec<_>>().join(", "),
            provider_reason
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_code_classification() {
        let result = classify_task("Write a Rust function to parse JSON");
        assert_eq!(result.category, TaskCategory::Code);
        assert_eq!(result.suggested_provider, CloudProvider::Claude);
    }

    #[test]
    fn test_research_classification() {
        let result = classify_task("What are the latest news about AI in 2025?");
        assert_eq!(result.category, TaskCategory::Research);
        assert_eq!(result.suggested_provider, CloudProvider::Perplexity);
    }

    #[test]
    fn test_creative_classification() {
        let result = classify_task("Write a short story about a robot");
        assert_eq!(result.category, TaskCategory::Creative);
        assert_eq!(result.suggested_provider, CloudProvider::ChatGPT);
    }

    #[test]
    fn test_image_classification() {
        let result = classify_task("Generate an image of a sunset over mountains");
        assert_eq!(result.category, TaskCategory::Image);
        assert_eq!(result.suggested_provider, CloudProvider::ChatGPT);
    }
}
